theano/tensor/alt_gemm_template.c

```diff
@@ -0,0 +1,133 @@
+/** %(name)s **/
+void alt_numpy_scalar_matrix_product_in_place_%(float_type)s(%(float_type)s scalar, PyArrayObject* matrix) {
+    NpyIter* iterator = NpyIter_New(matrix, 
+        NPY_ITER_READWRITE | NPY_ITER_EXTERNAL_LOOP | NPY_ITER_REFS_OK, 
+        NPY_KEEPORDER, NPY_NO_CASTING, NULL);
+    if(iterator == NULL)
+        alt_fatal_error("Unable to iterate over a matrix "
+                        "for a scalar * matrix operation.");
+    NpyIter_IterNextFunc* get_next = NpyIter_GetIterNext(iterator, NULL);
+    char** data_ptr = NpyIter_GetDataPtrArray(iterator);
+    npy_intp* stride_ptr = NpyIter_GetInnerStrideArray(iterator);
+    npy_intp* innersize_ptr = NpyIter_GetInnerLoopSizePtr(iterator);
+    do {
+        char* data = *data_ptr;
+        npy_intp stride = *stride_ptr;
+        npy_intp count = *innersize_ptr;
+        while(count) {
+            *((%(float_type)s*)data) *= scalar;
+            data += stride;
+            --count;
+        }
+    } while(get_next(iterator));
+    NpyIter_Deallocate(iterator);
+}
+/*Matrix+Matrix function. Compute (coeffA * matrixA) + (coeffB * matrixB)
+ * Remark: This function actually sums a C-contiguous matrix (alpha*op(A)*op(B)) with a F-contiguous matrix (beta*C)
```

Comment:
OK ! Sorry for the confusion, I will rewrite the comment and this function's parameters names so that there will be no more confusion with the gemm parameters.

---
I just realized that my point about F-contiguous vs C-contiguous sum seems useless !

In fact, the problem come from gemm implementation when BETA != 0: PyArray_MatrixProduct() returns a C-contiguous matrix (that is actually op(A)*op(B)), while C matrix must be read as F-contiguous. So the sum function actually works with a C-contiguous matrix (his first parameter) and a F-contiguous matrix (second parameter), so that the iteration with NpyIter could be fast for one matrix but less for the other.

I initially thought that the iteration would be faster if both matrices have the same "contiguousity" because we could then iterate it simultaneously just through memory order, regardless of the logic order, making the iteration more faster, and that's why I had made this comment.

But in fact, it also depends on the strides of the matrices: if both matrices have not the same strides (for example, if LDC != M), we will still have to iterate in the "logic" way, and then there will not be real performance gain.

So I think I will remove this comment !